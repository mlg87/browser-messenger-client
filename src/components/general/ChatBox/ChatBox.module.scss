.container {
    display: flex;
    flex-direction: column;
    flex-grow: 2;
    height: 45%;
    padding: 10px 20px;

    header {
        margin: 10px;
    }
}

.conversation {
    border: 1px dashed black;
    display: flex;
    flex-direction: column;
    flex-grow: 1;
    margin: 10px;
    max-height: 600px;
    min-height: 300px;
    padding: 10px;
    overflow: scroll;
}

.messageBubble {
    border-radius: 2px;
    margin: 5px;
    padding: 5px 10px;

    .timestamp {
        font-size: 11px;
    }

    &.currentUser {
        align-self: flex-end;
        background-color: rgba(green, 0.3);
    }

    &.otherUser {
        align-self: flex-start;
        background-color: rgba(gray, 0.3);
    }
}

.messageContainer {
    display: flex;
    flex-grow: 1;
    margin: 10px;
    margin-top: 20px;

    .messageInput {
        flex-grow: 1;
        height: 40px;
    }

    .submit {
        min-width: 100px;
    }
}

.typingIndicator {
    height: 20px;
    margin-left: 10px;
}