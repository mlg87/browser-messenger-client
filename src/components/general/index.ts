export { default as ChatBox } from './ChatBox/ChatBox';
export { default as UserCard } from './UserCard/UserCard';
export { default as UserList } from './UserList/UserList';
