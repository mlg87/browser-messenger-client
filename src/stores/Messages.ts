import { action, observable } from 'mobx';
import { Api } from '../services';

export class Messages {

    //
    // ─── OBSERVABLES ────────────────────────────────────────────────────────────────
    //

    @observable conversationId?: number;
    @observable conversations: any[] = []; // TODO: type this
    @observable messages: any[] = []; // TODO: type this

    //
    // ─── ACTIONS ────────────────────────────────────────────────────────────────────
    //

    @action createConversation = async (selectedUserId: number) => {

        try {

            const res = await Api.post('/conversations', { selectedUserId });

            if (!res.ok) {

                const message = res.data && res.data.message;
                throw Error(message || 'Error creating conversation.');

            }

            this.conversations.push(res.data.conversation);
            this.conversationId = res.data.conversation.id;

        } catch (error) {

            return Promise.reject(error.message);

        }

    }

    @action getMessagesForConversation = async () => {

        try {

            // REVIEW: perhaps drop the messages
            const res = await Api.get(`/conversations/${this.conversationId}/messages`, null);

            if (!res.ok) {

                const message = res.data && res.data.message;
                throw Error(message || 'Error fetching messages for conversation.');

            }

            this.messages = res.data.messages;

        } catch (error) {

            return Promise.reject(error.message);

        }

    }

    @action fetchConversations = async () => {

        try {

            const res = await Api.get('/conversations', null);

            if (!res.ok) {

                const message = res.data && res.data.message;
                throw Error(message || 'Error fetching conversations.');

            }

            this.conversations = res.data.conversations;

        } catch (error) {

            this.conversations = [];
            return Promise.reject(error.message);

        }

    }

}

export default new Messages();
