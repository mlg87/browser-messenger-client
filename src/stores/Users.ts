import { action, observable } from 'mobx';
import { Api } from '../services';
import { IUser } from './Auth/types';

export class Users {

    //
    // ─── OBSERVABLES ────────────────────────────────────────────────────────────────
    //

    @observable selectedUser?: IUser;
    @observable totalCount: number = 0;
    @observable users: IUser[] = [];

    //
    // ─── ACTIONS ────────────────────────────────────────────────────────────────────
    //

    @action fetchUsers = async () => {

        try {

            const res = await Api.get('/users', null);

            if (!res.ok) {

                const message = res.data && res.data.message;
                throw Error(message || 'Error fetching users.');

            }

            this.totalCount = res.data.count;
            this.users = res.data.users;

            return Promise.resolve();

        } catch (error) {

            this.users = [];

            return Promise.reject(error.message);

        }

    }

    @action updateUsersList = (user: IUser) => {

        const index = this.users.findIndex(u => u.id === user.id);

        this.users[index] = user;

    }

}

const usersStore = new Users();
export default usersStore;
