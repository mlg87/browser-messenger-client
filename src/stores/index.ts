import AuthStore, { Auth } from './Auth/Auth';
import MessagesStore, { Messages } from './Messages';
import UsersStore, { Users } from './Users';

export default {
    auth: AuthStore,
    messages: MessagesStore,
    users: UsersStore
};

export interface IStores {
    auth: Auth;
    messages: Messages;
    users: Users;
}
