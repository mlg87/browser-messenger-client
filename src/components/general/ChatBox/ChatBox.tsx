import { toJS } from 'mobx';
import { observer } from 'mobx-react';
import moment from 'moment';
import React, { useEffect, useRef, useState } from 'react';
import io from 'socket.io-client';
import genStyles from '../../../common/styles/general.module.scss';
import useDebounce from '../../../common/utils/useDebounce';
import { withStores } from '../../../common/utils/withStores';
import { Auth } from '../../../stores/Auth/Auth';
import { IUser } from '../../../stores/Auth/types';
import { Messages } from '../../../stores/Messages';
import { Users } from '../../../stores/Users';
import styles from './ChatBox.module.scss';

const URL = process.env.REACT_APP_API_URL || 'http://localhost:5000';

const ChatBox: React.FunctionComponent<IChatBoxProps> = props => {

    const conversationRef = useRef<HTMLDivElement>(null);
    const { current: socket } = useRef(io(URL, {
        query: {
            id: props.auth.user!.id
        }
    }));
    // needs to be destructured in main render
    const { selectedUser } = props.users;
    const { conversationId } = props.messages;

    // used to cleanup values of form when the selected user changes
    // tslint:disable-next-line:max-line-length
    const [currentSelectedUserId, setCurrentSelectedUserId] = useState<number | undefined>(selectedUser ? selectedUser.id : undefined);
    const [isFetching, setIsFetching] = useState(false);
    const [isOtherUserTyping, setIsOtherUserTyping] = useState(false);
    const [message, setMessage] = useState<string>('');

    const state: IChatBoxState = {
        currentSelectedUserId,
        isFetching,
        isOtherUserTyping,
        message,
        setCurrentSelectedUserId,
        setIsFetching,
        setIsOtherUserTyping,
        setMessage
    };

    // clears form when a new user is selected to chat with
    useCleanup(selectedUser, state, props);

    // update typing indicator on server
    const debouncedMessage = useDebounce(message, 250);
    useEffect(() => {

        let isTyping = false;
        if (debouncedMessage) {

            isTyping = true;

        }

        socket.emit('user-typing', {
            conversationId,
            isTyping,
            userId: props.auth.user!.id
        });

    }, [debouncedMessage]);

    // socket config
    useEffect(() => {

        socket.on('user-connectivity', user => {

            props.users.updateUsersList(user);

        });

        socket.on('update-conversation', newMessage => {

            props.messages.messages.push(newMessage);
            _scrollToBottom(conversationRef);

        });

        socket.on('update-user-typing', event => {

            const { isTyping, userId } = event;
            if (userId === props.users.selectedUser!.id) {

                setIsOtherUserTyping(isTyping);

            }

        });

        socket.on('error', event => {

            // TODO: handle this

        });

        return () => {

            socket.close();

        };

    }, []);

    // get or create the conversation with the selected user
    useEffect(() => {

        if (selectedUser) {

            const conversation = toJS(props.messages.conversations).find(c => {

                return c.user1.id === selectedUser.id || c.user2.id === selectedUser.id;

            });

            if (!conversation) {

                props.messages.createConversation(selectedUser.id);

            } else {

                props.messages.conversationId = conversation.id;

            }

        }

    }, [selectedUser]);

    useEffect(() => {

        if (conversationId) {

            setIsFetching(true);
            props.messages.getMessagesForConversation().then(() => setIsFetching(false));
            socket.emit('join-room', conversationId);

        }

    }, [conversationId]);

    useEffect(() => {

        if (props.messages.messages.length) {

            _scrollToBottom(conversationRef);

        }

    }, [props.messages.messages]);

    return (
        <div className={styles.container}>
            <header>
                <h2 className={genStyles.sectionTitle}>selected conversation</h2>
                {_renderDescriptor(props)}
            </header>
            <div className={styles.conversation} ref={conversationRef}>
                {_renderConversation(state, props)}
            </div>
            {_renderTypingIndicator(state, props)}
            <form
                className={styles.messageContainer}
                data-testid='chat-box-message-form'
                onSubmit={e => _handleSubmit(e, socket, state, props)}
            >
                <input
                    className={styles.messageInput}
                    disabled={!selectedUser}
                    onChange={e => _handleChange(e, state)}
                    placeholder={selectedUser ? 'start typing...' : 'select a user to start a conversation'}
                    value={message}
                />
                <input
                    className={styles.submit}
                    disabled={!selectedUser || !message}
                    type='submit'
                    value='send'
                />
            </form>
        </div>
    );

};

//
// ─── HOOKS ──────────────────────────────────────────────────────────────────────
//

const useCleanup = (
    selectedUser: IUser | undefined,
    state: IChatBoxState,
    props: IChatBoxProps
) => {

    useEffect(() => {

        if (selectedUser && selectedUser.id !== state.currentSelectedUserId) {

            state.setMessage('');
            state.setCurrentSelectedUserId(selectedUser.id);
            props.messages.messages = [];

        }

    }, [selectedUser]);

};

//
// ─── RENDERING ──────────────────────────────────────────────────────────────────
//

const _renderConversation = (
    state: IChatBoxState,
    props: IChatBoxProps
) => {

    const { messages, auth } = props;

    if (state.isFetching) {

        return <div>loading...</div>;

    }

    if (!messages.messages.length && props.users.selectedUser) {

        return <div>this is the beginning of your conversation</div>;

    }

    return toJS(messages.messages).map(m => {

        const classes = [styles.messageBubble];
        // messages returned from socket dont have full user as from, just id
        if (auth.user!.id === (m.from.id || m.from)) {

            classes.push(styles.currentUser);

        } else {

            classes.push(styles.otherUser);

        }

        return (
            <div key={m.id} className={classes.join(' ')}>
                <div className={styles.timestamp}>{moment.utc(m.createdAt).local().format('HH:mm MM/DD/YY')}</div>
                <div>{m.message}</div>
            </div>
        );

    });

};

const _renderDescriptor = (
    props: IChatBoxProps
) => {

    const { selectedUser } = props.users;

    if (!selectedUser) {

        return <i>no conversation selected. pick one from the user list</i>;

    }

    return <i>conversation with {selectedUser.username}</i>;

};

const _renderTypingIndicator = (
    state: IChatBoxState,
    props: IChatBoxProps
) => {

    let content = '';

    if (state.isOtherUserTyping) {

        content = `${props.users.selectedUser!.username} is typing...`;

    }

    return <div className={styles.typingIndicator}>{content}</div>;

};

//
// ─── HANDLERS ───────────────────────────────────────────────────────────────────
//

export const _handleChange = (
    e,
    state: IChatBoxState
) => {

    state.setMessage(e.target.value);

};

export const _handleSubmit = (
    e: React.SyntheticEvent,
    socket: SocketIOClient.Socket,
    state: IChatBoxState,
    props: IChatBoxProps
) => {

    e.preventDefault();

    // ts null check
    if (!socket) {

        return null;

    }

    socket.emit('new-message', {
        conversationId: props.messages.conversationId,
        from: props.auth.user!.id,
        to: props.users.selectedUser!.id,
        message: state.message
    });

    state.setMessage('');

};

//
// ─── HELPERS ────────────────────────────────────────────────────────────────────
//

const _scrollToBottom = (
    ref: React.RefObject<HTMLDivElement>
) => {

    if (!ref.current) {

        return null;

    }

    ref.current.scrollTop = ref.current.scrollHeight;

};

export default withStores('auth', 'messages', 'users')(observer(ChatBox));

//
// ─── INTERFACES ─────────────────────────────────────────────────────────────────
//

export interface IChatBoxProps {
    readonly auth: Auth;
    readonly messages: Messages;
    readonly users: Users;
}

export interface IChatBoxState {
    readonly currentSelectedUserId?: number;
    readonly isFetching: boolean;
    readonly isOtherUserTyping: boolean;
    readonly message?: string;
    readonly setCurrentSelectedUserId: (value?: number) => void;
    readonly setIsFetching: (value: boolean) => void;
    readonly setIsOtherUserTyping: (value: boolean) => void;
    readonly setMessage: (value: string) => void;
}
