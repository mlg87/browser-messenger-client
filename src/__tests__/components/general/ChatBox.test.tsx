import { act, fireEvent, render } from '@testing-library/react';
import React from 'react';
import io from 'socket.io-client';
import { ChatBox } from '../../../components/general';
import { IChatBoxProps } from '../../../components/general/ChatBox/ChatBox';
import appState from '../../../stores';
import { Messages } from '../../../stores/Messages';

jest.mock('socket.io-client', () => {

    const socket = {
        close: jest.fn(),
        emit: jest.fn(),
        on: jest.fn(),

    };

    return () => socket;

});

const startingProps: IChatBoxProps = {
    ...appState,
    auth: {
        ...appState.auth,
        user: {
            id: 23,
            isOnline: true,
            username: 'test-auth-user'
        }
    }
};

const setup = (props: IChatBoxProps = startingProps) => render(

    <ChatBox {...props} />
);

describe('ChatBox', () => {

    beforeEach(() => {

        jest.resetAllMocks();

    });

    it('renders', () => {

        const { container } = setup();
        expect(container.firstChild).toMatchSnapshot();

    });

    describe('the conversation between the authed user and another user', () => {

        it('renders a message saying it is the beginning of the convesation if there are no messages', () => {

            const props = {
                ...startingProps,
                users: {
                    ...startingProps.users,
                    selectedUser: {
                        id: 456,
                        isOnline: false,
                        username: 'test-456'
                    }
                }
            };
            const { container } = setup(props);

            expect(container.firstChild).toMatchSnapshot();

        });

        it('renders the messages from the users if there are any', () => {

            const messages = new Messages();
            messages.messages = [
                {
                    createdAt: '2019-07-02 21:51:29.637306', from: 23, id: 1, to: 456, message: 'hi'
                }
            ];

            const props = {
                ...startingProps,
                users: {
                    ...startingProps.users,
                    selectedUser: {
                        id: 456,
                        isOnline: false,
                        username: 'test-456'
                    }
                },
                messages
            };
            const { container } = setup(props);

            expect(container.firstChild).toMatchSnapshot();

        });

        it('submits the message when the user pushes send or hits enter and there is a value in the input', () => {

            const conversationId = 999;
            const messages = new Messages();
            messages.conversationId = conversationId;
            messages.messages = [
                {
                    createdAt: '2019-07-02 21:51:29.637306', from: 23, id: 1, to: 456, message: 'hi'
                }
            ];

            const props = {
                ...startingProps,
                users: {
                    ...startingProps.users,
                    selectedUser: {
                        id: 456,
                        isOnline: false,
                        username: 'test-456'
                    }
                },
                messages
            };
            const message = 'second message';
            const { getByPlaceholderText, getByTestId } = setup(props);
            const mocket = io();

            fireEvent.change(getByPlaceholderText('start typing...'), { target: { value: message } });
            fireEvent.submit(getByTestId('chat-box-message-form'));

            expect(mocket.emit).toHaveBeenLastCalledWith('new-message', {
                conversationId,
                from: startingProps.auth.user!.id,
                message,
                to: props.users.selectedUser.id
            });

        });

    });

});
