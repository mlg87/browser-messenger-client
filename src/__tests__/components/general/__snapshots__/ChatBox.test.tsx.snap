// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatBox renders 1`] = `
<div
  class="container"
>
  <header>
    <h2
      class="sectionTitle"
    >
      selected conversation
    </h2>
    <i>
      no conversation selected. pick one from the user list
    </i>
  </header>
  <div
    class="conversation"
  />
  <div
    class="typingIndicator"
  />
  <form
    class="messageContainer"
    data-testid="chat-box-message-form"
  >
    <input
      class="messageInput"
      disabled=""
      placeholder="select a user to start a conversation"
      value=""
    />
    <input
      class="submit"
      disabled=""
      type="submit"
      value="send"
    />
  </form>
</div>
`;

exports[`ChatBox the conversation between the authed user and another user renders a message saying it is the beginning of the convesation if there are no messages 1`] = `
<div
  class="container"
>
  <header>
    <h2
      class="sectionTitle"
    >
      selected conversation
    </h2>
    <i>
      conversation with 
      test-456
    </i>
  </header>
  <div
    class="conversation"
  >
    <div>
      this is the beginning of your conversation
    </div>
  </div>
  <div
    class="typingIndicator"
  />
  <form
    class="messageContainer"
    data-testid="chat-box-message-form"
  >
    <input
      class="messageInput"
      placeholder="start typing..."
      value=""
    />
    <input
      class="submit"
      disabled=""
      type="submit"
      value="send"
    />
  </form>
</div>
`;

exports[`ChatBox the conversation between the authed user and another user renders the messages from the users if there are any 1`] = `
<div
  class="container"
>
  <header>
    <h2
      class="sectionTitle"
    >
      selected conversation
    </h2>
    <i>
      conversation with 
      test-456
    </i>
  </header>
  <div
    class="conversation"
  >
    <div
      class="messageBubble currentUser"
    >
      <div
        class="timestamp"
      >
        15:51 07/02/19
      </div>
      <div>
        hi
      </div>
    </div>
  </div>
  <div
    class="typingIndicator"
  />
  <form
    class="messageContainer"
    data-testid="chat-box-message-form"
  >
    <input
      class="messageInput"
      placeholder="start typing..."
      value=""
    />
    <input
      class="submit"
      disabled=""
      type="submit"
      value="send"
    />
  </form>
</div>
`;
